### PR TITLE
prov/verbs: Optimize verbs critical path

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
This PR applies the following optimization for the verbs' critical path:
- Drop excessive `ibv_send|recv_wr` structures from WR entry -- allows us to avoid `memset()` call
- Optimize CQ entry processing by refactoring completion flags handling

### CODE REVIEW

**Comment by shefty** [comment]
do not merge until after rc1

File: prov/verbs/src/verbs_cq.c [hunk 1]
```diff
@@ -36,39 +36,46 @@
...
+			*data = ntohl(wc->imm_data);
+		return FI_REMOTE_CQ_DATA;
+	} else {
+		if (data)
+			*data = 0;
+		return 0;
+	}
+}
 
+static inline uint64_t fi_ibv_comp_flags(struct ibv_wc *wc, size_t *len)
```
Comment (shefty) [inline prov/verbs/src/verbs_cq.c]: Please rename to something like fi_ibv_handle_wc(wc, &flags, &len, &data) and merge this function with fi_ibv_handle_imm_data() above.  This should provide another level of optimization, since immediate data only applies to IBV_WC_RECV_RDMA_WITH_IMM or IBV_WC_RECV opcodes.  Since 3 different fields are being set -- flags, data, and len -- I would just pass all into the function.

File: prov/verbs/src/verbs_cq.c [hunk 2]
```diff
@@ -302,7 +302,7 @@ static inline int fi_ibv_poll_outstanding_cq(struct fi_ibv_msg_ep *ep,
 			    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
 				ret = fi_ibv_wc_2_wce(cq, &wc, &wce);
 				if (ret) {
-					wre->ep = NULL;
+					wre->srq = NULL;
```
Comment (shefty) [inline prov/verbs/src/verbs_cq.c]: This change should go into a separate commit

File: prov/verbs/src/verbs_cq.c [hunk 3]
```diff
@@ -405,29 +405,26 @@ ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
 	    ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)) {
 		wre = (struct fi_ibv_wre *)(uintptr_t)wc->wr_id;
 		assert(wre && (wre->ep || wre->srq));
+		wc->wr_id = (uintptr_t)wre->context;
+
+		if (wc->status == IBV_WC_WR_FLUSH_ERR)
+			/* Handles case where remote side destroys
+			 * the connection, but local side isn't aware
+			 * about that yet */
+			ret = 0;
```
Comment (shefty) [inline prov/verbs/src/verbs_cq.c]: nit: Move comment above if statement or add { }'s

File: prov/verbs/src/verbs_cq.c (hunk 1 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_cq.c]: Thanks! I like your suggestion. Done

File: prov/verbs/src/verbs_cq.c (hunk 3 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_cq.c]: Done

File: prov/verbs/src/verbs_cq.c (hunk 2 above)
Comment (gladkovdmitry17) [reply prov/verbs/src/verbs_cq.c]: It should be dropped from this PR

**Comment by gladkovdmitry17** [comment]
The PR has been updated in accordance with the comments provided. Merging this to not stop upcoming work. Thanks!
